internal/controllers/machine/machine_controller_phases.go

```diff
@@ -116,9 +116,8 @@ func (r *Reconciler) reconcileExternal(ctx context.Context, cluster *clusterv1.C
 		return external.ReconcileOutput{}, err
 	}
 
-	// With the migration from v1alpha2 to v1alpha3, Machine controllers should be the owner for the
-	// infra Machines, hence remove any existing machineset controller owner reference
-	if controller := metav1.GetControllerOf(obj); controller != nil && controller.Kind == "MachineSet" {
+	// Ensure the Machine is the controller owner for infrastructure and bootstrap object it references.
+	if controller := metav1.GetControllerOf(obj); controller != nil && controller.Kind != "Machine" {
```

Comment:
I'm not sure that works as expected:

CP machine:
```yaml
  ownerReferences:
  - apiVersion: controlplane.cluster.x-k8s.io/v1beta1
    kind: KubeadmControlPlane
    name: quick-start-56f9nk-4qsdk
    uid: cd5bf608-32d6-4eca-9394-d8699c9ebd29
  - apiVersion: cluster.x-k8s.io/v1beta1
    blockOwnerDeletion: true
    controller: true
    kind: Machine
    name: quick-start-56f9nk-4qsdk-w27r4
    uid: 348a1f04-098b-43c5-b77b-faa0b81e2ff2
```

MS machine:
```yaml
  ownerReferences:
  - apiVersion: cluster.x-k8s.io/v1beta1
    kind: MachineSet
    name: quick-start-56f9nk-md-0-vnk7j-d4cfd44c8
    uid: 9ea10555-bf2b-46b5-89bf-139cb777919b
  - apiVersion: cluster.x-k8s.io/v1beta1
    blockOwnerDeletion: true
    controller: true
    kind: Machine
    name: quick-start-56f9nk-md-0-vnk7j-d4cfd44c8-k78dv
    uid: d9ef3464-797a-4777-aa99-01f80f921a25
```

(the MS / KCP ownerRefs don't have `controller:true` and `GetControllerOf` only returns those.

Also otherwise I would wonder if we depend on ordering, given that `GetControllerOf` returns the first controller ownerRef (not sure if there can be multiple)

Reply:
Apologies - should have added WIP to this - I'm still working on it 😄

Reply:
Ah got it, np :)